### PR TITLE
[MRG] Simulate upload from local

### PR DESCRIPTION
**Pull request by gtdang**
Creates the ability to use local files for simulating GUI file uploads for unit-testing. Previously there was only the ability to load from URL sources. This will allow the use of local files so developers are not required an internet connection to run tests and to make use of default configuration files that are included in the repo.

Extracted from #808

### CODE REVIEW

**Comment by ntolley** [comment]
I think there might need to be some logic for setting up windows specific file paths:
image

**Comment by kmilo9999** [comment]
> I think there might need to be some logic for setting up windows specific file paths: image


I have something similar in test_gui.py:621  where the path is modified for windows systems. Maybe it works for you too.

**Comment by gtdang** [comment]
@ntolley @kmilo9999 
Ok finally got the checks to pass. There were two issues to get them working on Windows. 

1. The code first tries to load whatever path you give it (URL or local path) locally, if that doesn't work it tries to use it as a URL. It does this by catching exceptions in a try/except clause. Giving a URL to a the `open` function on Unix will result in a `FileNotFound` error but on Windows it will be a `OSError` because the  `/` characters in URLs. Basically I just needed to also catch the `OSError`. 
2. I was doing asserts with file sizes. I learned that text files saved to disk on Windows are larger than Unix text files because their carriage returns/next line characters are larger.  This will result in a larger file on disk than if stream directly to memory from a URL. I changed the asserts to check if the files are decoded to a dictionary the same way.

File: hnn_core/gui/gui.py [hunk 1]
```diff
@@ -836,21 +837,55 @@ def _simulate_viz_action(self, action_name, *args, **kwargs):
...
     }]
 
+    return upload_structure
+
+
+def _prepare_upload_file(path):
+    """ Simulates output of the FileUpload widget for testing.
+
+    Unit tests for the GUI simulate user upload of files. File source can
+    either be locally or from a URL. This function returns the data structure
```
Comment (ntolley) [inline hnn_core/gui/gui.py:878]:
```suggestion
    either be local or from a URL. This function returns the data structure
```

File: hnn_core/gui/gui.py [hunk 2]
```diff
@@ -836,21 +837,55 @@ def _simulate_viz_action(self, action_name, *args, **kwargs):
         action(*args, **kwargs)
 
 
+def _prepare_upload_file_from_local(path):
+    with open(path, 'rb') as file:
+        content = memoryview(file.read())
+    last_modified = datetime.fromtimestamp(path.stat().st_mtime)
+
+    upload_structure = [{
+        'name': path.name,
```
Comment (ntolley) [inline hnn_core/gui/gui.py:846]: Is there any significance to using the full path for `name` in the local file vs. that last characters after a `'/'` in the URL upload? Is it just metadata so it doesn't matter that much?

File: hnn_core/tests/test_gui.py [hunk 3]
```diff
@@ -50,6 +54,36 @@ def test_gui_load_params():
...
+    assert (content_from_url['name'] ==
+            content_from_local['name'] ==
+            'default.json')
+    assert (content_from_url['type'] ==
+            content_from_local['type'] ==
+            'application/json')
+    # Check that the size attribute is present. Cannot do an equivalency check
+    # because file systems may add additional when saving to disk.
+    assert content_from_url.get('size')
+    assert content_from_local.get('size')
```
Comment (ntolley) [inline hnn_core/tests/test_gui.py:79]:
```suggestion
    assert 'size' in content_from_url
    assert 'size' in content_from_local
```
Little more readable

File: hnn_core/gui/gui.py (hunk 2 above)
Comment (gtdang) [reply hnn_core/gui/gui.py:847]: The name attribute of a Path object is the filename. Though this reminds me that all to add a conversion to Path in this function for if a string path is passed as an argument.

**Review by ntolley — APPROVED** [review]
All good on my end! merging now since this is pretty compact/self-contained

**Pull request merged**